sleepecg/io/sleep_readers.py

```diff
@@ -0,0 +1,269 @@
+# Authors: Florian Hofer
+#
+# License: BSD (3-clause)
+
+"""Read datasets containing ECG data and sleep stage annotations."""
+
+import datetime
+from dataclasses import dataclass
+from enum import IntEnum
+from pathlib import Path
+from typing import Iterator, NamedTuple, Optional, Union
+from xml.etree import ElementTree
+
+import numpy as np
+
+from ..config import get_config
+from ..heartbeats import detect_heartbeats
+from .nsrr import get_nsrr_url, list_nsrr
+from .utils import download_file
+
+__all__ = [
+    'SleepRecord',
+    'SleepStage',
+    'read_mesa',
+]
+
+
+class SleepStage(IntEnum):
+    """
+    Mapping of AASM sleep stages to integers.
+
+    To facilitate hypnogram plotting, values increase with wakefulness.
+    """
+
+    WAKE = 5
+    REM = 4
+    N1 = 3
+    N2 = 2
+    N3 = 1
+    UNDEFINED = -1
+
+
+@dataclass
+class SleepRecord:
+    """
+    Dataclass to store a single sleep record.
+
+    Attributes
+    ----------
+    sleep_stages : np.ndarray, optional
+        Sleep stages according to AASM guidelines, stored as integers as
+        defined by `SleepStage`, by default `None`
+    sleep_stage_duration : int, optional
+        Duration of each sleep stage in seconds, by default `None`
+    id : str, optional
+        The record's ID, by default `None`.
+    recording_start_time : datetime.time, optional
+        Time at which the recording was started, by default `None`.
+    heartbeat_times : np.ndarray, optional
+        Times of heartbeats relative to recording start in seconds, by
+        default `None`.
+    """
+
+    sleep_stages: Optional[np.ndarray] = None
+    sleep_stage_duration: Optional[int] = None
+    id: Optional[str] = None
+    recording_start_time: Optional[datetime.time] = None
+    heartbeat_times: Optional[np.ndarray] = None
+
+
+class _ParseNsrrXmlResult(NamedTuple):
+    sleep_stages: np.ndarray
+    sleep_stage_duration: int
+    recording_start_time: datetime.time
+
+
+def _parse_nsrr_xml(xml_filepath: Path) -> _ParseNsrrXmlResult:
+    """
+    Parse NSRR xml sleep stage annotation file.
+
+    Parameters
+    ----------
+    xml_filepath : pathlib.Path
+        Path of the annotation file to read.
+
+    Returns
+    -------
+    sleep_stages : np.ndarray
+        Sleep stages according to AASM guidelines, stored as integers as
+        defined by `SleepStage`.
+    fs_sleep_stages : float
+        Sampling frequency of `sleep_stages`.
```

Comment:
This should be `sleep_stage_duration: int` I guess.

Reply:
```suggestion
    sleep_stage_duration : int
        Duration of each sleep stage in seconds.
```